README.md
# animated_dropdown_button

Simple dropdown button with an interesting animation.
Currently it can only list Strings, whose value can be retrieved by the AnimatedDropdownButtonController.
Use without moderation.

## Getting Started

You start by instantiating an AnimatedDropdownButtonController with a string list and an initial value, which must be present in the list.
Then you can instantiate an AnimatedDropdownButton with the created controller.

Now you can use an AnimatedSearchBox. Just instantiate an AnimatedSearchBoxController first.
AnimatedSearchBox is fully customizable.


CHANGELOG.md
## 1.1.5

* Added AnimatedSearchBox widget.

## 1.0.5

* Fixed arrow animation.

## 1.0.0

* First release.


example/lib/main.dart
import 'package:animated_dropdown_button/animated_dropdown_button.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Animated Dropdown Button Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'ANIMATED DROPDOWN BUTTON DEMO'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class AnimalModel extends AnimatedDropdownItem {
  final String name;

  AnimalModel(this.name);

  @override
  String get text => name;

  @override
  String get value => name;
}

class _MyHomePageState extends State<MyHomePage> {
  late final AnimatedSearchBoxController animatedSearchController;

  final AnimatedDropdownButtonController animalsController = AnimatedDropdownButtonController(
    items: [
      'Cat',
      'Dog',
      'Bird',
      'Cat',
      'Dog',
    ],
    initialValue: 'Dog',
  );

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    animatedSearchController = AnimatedSearchBoxController(
      items: animals,
      initialValue: AnimalModel(
        'Cat',
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      // resizeToAvoidBottomInset: false,
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: Container(
        color: const Color.fromARGB(255, 165, 149, 149),
        child: Center(
          child: SingleChildScrollView(
            child: Column(
              children: <Widget>[
                const SizedBox(height: 50),
                AnimatedSearchBox(
                  width: 336,
                  controller: animatedSearchController,
                  backgroundColor: Colors.white,
                  hintText: 'Type an animal name',
                ),
                const SizedBox(
                  height: 50,
                ),
                AnimatedDropdownButton(
                  controller: animalsController,
                  backgroundColor: Colors.red,
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}

final List<AnimalModel> animals = [
  AnimalModel(
    'Cat',
  ),
  AnimalModel(
    'Dog',
  ),
  AnimalModel(
    'Bird',
  ),
  AnimalModel(
    'Donkey',
  ),
  AnimalModel(
    'Moneky',
  ),
  AnimalModel(
    'Cabra',
  ),
  AnimalModel(
    'Cavalo',
  ),
  AnimalModel(
    'Leão',
  ),
];


lib/src/animated_dropdown_button_widget.dart
library animated_dropdown_button;

import 'package:flutter/material.dart';

class AnimatedDropdownButton extends StatelessWidget {
  const AnimatedDropdownButton({
    super.key,
    required this.controller,
    this.backgroundColor,
    this.innerTextStyle,
  });

  final AnimatedDropdownButtonController controller;
  final Color? backgroundColor;
  final TextStyle? innerTextStyle;

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        ClipRect(
          child: Container(
            width: 175,
            alignment: const Alignment(0, -1),
            child: _AnimatedDropdown(
              controller: controller,
              backgroundColor: backgroundColor,
              innerTextStyle: innerTextStyle,
            ),
          ),
        ),
      ],
    );
  }
}

class _AnimatedDropdown extends StatefulWidget {
  const _AnimatedDropdown({
    required this.controller,
    this.backgroundColor,
    this.innerTextStyle,
  });

  @override
  State<_AnimatedDropdown> createState() => __AnimatedDropdownState();
  final AnimatedDropdownButtonController controller;
  final Color? backgroundColor;
  final TextStyle? innerTextStyle;
}

class __AnimatedDropdownState extends State<_AnimatedDropdown> with SingleTickerProviderStateMixin {
  bool _isExpanded = false;
  late AnimationController _animationController;
  late Animation<double> _animation;
  OverlayEntry? entry;

  @override
  void initState() {
    super.initState();
    _animationController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 500),
    );
    _animation = CurvedAnimation(
      parent: _animationController,
      curve: Curves.elasticInOut,
    );
    WidgetsBinding.instance.addPostFrameCallback((_) {});
  }

  void showOverlay() {
    final overlay = Overlay.of(context);
    final renderBox = context.findRenderObject() as RenderBox;
    final offset = renderBox.localToGlobal(Offset.zero);

    entry = OverlayEntry(
      builder: (context) => Positioned(
        height: 1000,
        width: 800,
        child: Stack(
          children: [
            _DropdownOverlay(onDismiss: _toggleExpand),
            Positioned(
              left: offset.dx,
              top: offset.dy,
              child: Material(
                borderOnForeground: false,
                color: Colors.transparent,
                child: AnimatedContainer(
                  curve: _isExpanded ? Curves.elasticOut : Curves.easeInOut,
                  duration: _isExpanded ? const Duration(seconds: 1) : const Duration(milliseconds: 400),
                  height: _isExpanded
                      ? ((42 + (15 / widget.controller.items.length)) * widget.controller.items.length.toDouble())
                      : 40,
                  child: buildOverlay(),
                ),
              ),
            ),
          ],
        ),
      ),
    );

    overlay.insert(entry!);
  }

  Widget buildOverlay() {
    return GestureDetector(
      onTap: _toggleExpand,
      child: Container(
        alignment: Alignment.topCenter,
        width: 175,
        decoration: BoxDecoration(
          color: widget.backgroundColor ?? const Color.fromARGB(255, 164, 28, 255),
          borderRadius: const BorderRadius.all(
            Radius.circular(30),
          ),
        ),
        child: ListView(
          padding: EdgeInsets.zero,
          physics: const NeverScrollableScrollPhysics(),
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                _selectedValue(),
                Padding(
                  padding: const EdgeInsets.only(right: 16.0, top: 8.0),
                  child: RotationTransition(
                    turns: Tween(begin: 0.0, end: 0.5).animate(_animation),
                    child: const Icon(Icons.keyboard_arrow_down, color: Colors.white),
                  ),
                ),
              ],
            ),
            _isExpanded ? const SizedBox(height: 4) : const SizedBox.shrink(),
            for (var i = 0; i < widget.controller.items.length; i++)
              GestureDetector(
                onTap: () {
                  setState(() {
                    widget.controller.selectedValue = widget.controller.items[i];
                  });
                  _toggleExpand();
                },
                child: AnimatedContainer(
                  clipBehavior: Clip.none,
                  padding: EdgeInsets.only(top: _isExpanded ? 4 : 0, bottom: _isExpanded ? 4 : 0),
                  curve: Curves.linear,
                  duration: const Duration(milliseconds: 300),
                  height: _isExpanded ? 34 : 0,
                  alignment: Alignment.centerLeft,
                  child: Padding(
                    padding: const EdgeInsets.only(left: 18.0),
                    child: Opacity(
                      opacity: _isExpanded ? 1 : 0,
                      child: Text(
                        widget.controller.items[i],
                        style: widget.innerTextStyle ?? const TextStyle(color: Colors.white),
                        textAlign: TextAlign.start,
                      ),
                    ),
                  ),
                ),
              ),
          ],
        ),
      ),
    );
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  void _toggleExpand() async {
    if (_isExpanded == false) {
      showOverlay();
      Future.delayed(const Duration(milliseconds: 200), () {
        setState(() {
          _isExpanded = true;
          entry!.markNeedsBuild();
        });
      });

      _animationController.forward();
    }

    if (_isExpanded == true) {
      _isExpanded = false;
      _animationController.reverse();
      entry!.markNeedsBuild();
      await Future.delayed(const Duration(milliseconds: 400));
      entry?.remove();
    }
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: _toggleExpand,
      child: Container(
        alignment: Alignment.topCenter,
        width: 175,
        height: 40,
        decoration: BoxDecoration(
          color: widget.backgroundColor ?? const Color.fromARGB(255, 164, 28, 255),
          borderRadius: const BorderRadius.all(
            Radius.circular(30),
          ),
        ),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            _selectedValue(),
            Padding(
              padding: const EdgeInsets.only(right: 16.0, top: 8.0),
              child: RotationTransition(
                turns: Tween(begin: 0.0, end: 0.5).animate(_animation),
                child: const Icon(Icons.keyboard_arrow_down, color: Colors.white),
              ),
            ),
          ],
        ),
      ),
    );
  }

  Widget _selectedValue() {
    return Padding(
      padding: const EdgeInsets.only(left: 18.0, top: 8.0),
      child: Text(
        widget.controller.selectedValue!,
        style: widget.innerTextStyle ?? const TextStyle(color: Colors.white),
        textAlign: TextAlign.start,
      ),
    );
  }
}

class _DropdownOverlay extends StatefulWidget {
  final VoidCallback onDismiss;

  const _DropdownOverlay({required this.onDismiss});

  @override
  State<_DropdownOverlay> createState() => __DropdownOverlayState();
}

class __DropdownOverlayState extends State<_DropdownOverlay> {
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: widget.onDismiss,
      child: Material(
        color: Colors.transparent,
        child: Container(
          color: Colors.transparent,
          height: MediaQuery.of(context).size.height,
          width: MediaQuery.of(context).size.width,
        ),
      ),
    );
  }
}

class AnimatedDropdownButtonController {
  final List<String> items;
  String? selectedValue;
  String initialValue;

  AnimatedDropdownButtonController({
    required this.items,
    required this.initialValue,
  })  : selectedValue = initialValue,
        assert(items.contains(initialValue)),
        assert(items.isNotEmpty),
        assert(items.length < 6);
}


lib/src/animated_search_box.dart
import 'package:flutter/material.dart';

class AnimatedSearchBox extends StatelessWidget {
  const AnimatedSearchBox({
    super.key,
    required this.controller,
    this.backgroundColor,
    this.itemsTextStyle,
    this.selectedTextStyle,
    this.width = 336,
    this.height = 36,
    this.hintText = '',
    this.hintTextStyle,
    this.descriptionWidget,
  });

  final AnimatedSearchBoxController controller;
  final Color? backgroundColor;
  final TextStyle? itemsTextStyle;
  final TextStyle? selectedTextStyle;
  final double width;
  final double height;
  final String hintText;
  final TextStyle? hintTextStyle;
  final Widget? descriptionWidget;

  @override
  Widget build(BuildContext context) {
    return Stack(
      children: [
        Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            descriptionWidget != null
                ? Padding(
                    padding: const EdgeInsets.only(bottom: 4.0),
                    child: descriptionWidget!,
                  )
                : const SizedBox.shrink(),
            ClipRect(
              child: Container(
                width: width,
                alignment: const Alignment(0, -1),
                child: _AnimatedSearchDropdown(
                  controller: controller,
                  backgroundColor: backgroundColor,
                  itemsTextStyle: itemsTextStyle,
                  selectedTextStyle: selectedTextStyle,
                  width: width,
                  hintText: hintText,
                  hintTextStyle: hintTextStyle,
                ),
              ),
            ),
          ],
        ),
      ],
    );
  }
}

class _AnimatedSearchDropdown extends StatefulWidget {
  const _AnimatedSearchDropdown({
    required this.controller,
    this.backgroundColor,
    required this.width,
    this.itemsTextStyle,
    this.selectedTextStyle,
    this.hintText = '',
    this.hintTextStyle,
  });

  @override
  State<_AnimatedSearchDropdown> createState() => _AnimatedSearchDropdownState();
  final AnimatedSearchBoxController controller;
  final Color? backgroundColor;
  final TextStyle? itemsTextStyle;
  final TextStyle? selectedTextStyle;
  final double width;
  final String hintText;
  final TextStyle? hintTextStyle;
}

class _AnimatedSearchDropdownState extends State<_AnimatedSearchDropdown> with SingleTickerProviderStateMixin {
  bool _isExpanded = false;
  bool _isInvisible = true;
  late AnimationController _animationController;
  late Animation<double> _animation;
  late FocusNode _focusNode;
  final ScrollController _scrollController = ScrollController();

  OverlayEntry? entry;
  final LayerLink _layerLink = LayerLink();
  GlobalKey globalKey = GlobalKey();

  @override
  void initState() {
    super.initState();
    _focusNode = FocusNode();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      globalKey;
    });
    _animationController = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 500),
    );
    _animation = CurvedAnimation(
      parent: _animationController,
      curve: Curves.elasticInOut,
    );

    widget.controller.initialize();

    _focusNode.addListener(() {
      if (_focusNode.hasFocus) {
        _toggleExpand();
      }
    });
  }

  void showOverlay() {
    final overlay = Overlay.of(context);
    final renderBox = context.findRenderObject() as RenderBox;
    final offset = renderBox.localToGlobal(Offset.zero);

    entry = OverlayEntry(
      builder: (context) => Positioned(
        child: Stack(
          children: [
            _DropdownOverlay(onDismiss: _toggleExpand),
            Positioned(
              left: offset.dx,
              top: offset.dy,
              child: CompositedTransformFollower(
                link: _layerLink,
                showWhenUnlinked: false,
                offset: Offset(0.0, renderBox.size.height),
                child: Material(
                  borderOnForeground: false,
                  color: Colors.transparent,
                  child: AnimatedOpacity(
                    opacity: _isInvisible ? 0 : 1,
                    duration: _isExpanded ? const Duration(milliseconds: 300) : const Duration(milliseconds: 300),
                    child: AnimatedContainer(
                      curve: _isExpanded ? Curves.elasticOut : Curves.easeInOut,
                      duration: _isExpanded ? const Duration(seconds: 1) : const Duration(milliseconds: 300),
                      height: _isExpanded ? 200 : 36,
                      width: renderBox.size.width,
                      child: buildOverlay(),
                    ),
                  ),
                ),
              ),
            ),
          ],
        ),
      ),
    );

    overlay.insert(entry!);
  }

  Widget buildOverlay() {
    return GestureDetector(
      onTap: _toggleExpand,
      child: Container(
          alignment: Alignment.topCenter,
          width: widget.width,
          decoration: BoxDecoration(
              color: widget.backgroundColor ?? const Color.fromARGB(255, 164, 28, 255),
              borderRadius: const BorderRadius.only(bottomLeft: Radius.circular(10), bottomRight: Radius.circular(10))),
          child: Column(
            children: [
              _isExpanded ? const SizedBox(height: 4) : const SizedBox.shrink(),
              _isExpanded
                  ? Expanded(
                      child: RawScrollbar(
                        thumbColor: const Color.fromARGB(255, 160, 158, 158),
                        padding: const EdgeInsets.only(right: 16),
                        thumbVisibility: true,
                        controller: _scrollController,
                        thickness: 3,
                        child: Theme(
                          data:
                              ThemeData(colorScheme: ColorScheme.fromSwatch().copyWith(secondary: Colors.transparent)),
                          child: ListView(
                            controller: _scrollController,
                            padding: EdgeInsets.zero,
                            children: [
                              for (var i = 0; i < widget.controller.filteredItems.length; i++)
                                GestureDetector(
                                  onTap: () {
                                    setState(() {
                                      widget.controller.selectedValue = widget.controller.filteredItems[i];

                                      if (widget.controller.selectedValue != null) {
                                        widget.controller.textEditingController.text =
                                            widget.controller.selectedValue!.text;
                                        _focusNode.unfocus();
                                      }
                                    });

                                    _toggleExpand();
                                  },
                                  child: AnimatedContainer(
                                    clipBehavior: Clip.none,
                                    padding: EdgeInsets.only(top: _isExpanded ? 4 : 0, bottom: _isExpanded ? 4 : 0),
                                    curve: Curves.linear,
                                    duration: const Duration(milliseconds: 300),
                                    height: _isExpanded ? 30 : 0,
                                    alignment: Alignment.centerLeft,
                                    child: Padding(
                                      padding: const EdgeInsets.only(left: 16.0),
                                      child: Opacity(
                                        opacity: _isExpanded ? 1 : 0,
                                        child: Text(
                                          widget.controller.filteredItems[i].text,
                                          style: widget.itemsTextStyle ?? const TextStyle(color: Colors.black),
                                          textAlign: TextAlign.start,
                                        ),
                                      ),
                                    ),
                                  ),
                                ),
                            ],
                          ),
                        ),
                      ),
                    )
                  : const SizedBox.shrink(),
            ],
          )),
    );
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  void _toggleExpand() async {
    if (_isExpanded == false) {
      _isInvisible = true;
      showOverlay();
      Future.delayed(const Duration(milliseconds: 200), () {
        setState(() {
          _isInvisible = false;
          _isExpanded = true;
          entry!.markNeedsBuild();
        });
      });
      _animationController.forward();
    } else {
      _isExpanded = false;
      _isInvisible = true;
      _animationController.reverse();
      entry!.markNeedsBuild();
      setState(() {});
      await Future.delayed(const Duration(milliseconds: 400));

      entry?.remove();

      _focusNode.unfocus();
    }
  }

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: _toggleExpand,
      child: CompositedTransformTarget(
        link: _layerLink,
        child: AnimatedContainer(
          duration: _isExpanded ? const Duration(milliseconds: 100) : const Duration(milliseconds: 500),
          alignment: Alignment.topCenter,
          width: widget.width,
          height: 36,
          decoration: BoxDecoration(
            color: widget.backgroundColor ?? const Color.fromARGB(255, 164, 28, 255),
            borderRadius: _isInvisible
                ? const BorderRadius.all(
                    Radius.circular(10),
                  )
                : const BorderRadius.only(topLeft: Radius.circular(10), topRight: Radius.circular(10)),
          ),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              _selectedValue(),
              Padding(
                padding: const EdgeInsets.only(right: 8.0),
                child: RotationTransition(
                  turns: Tween(begin: 0.0, end: 0.5).animate(_animation),
                  child: const Icon(Icons.keyboard_arrow_down, color: Colors.grey),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _selectedValue() {
    return AnimatedContainer(
      alignment: Alignment.center,
      duration: const Duration(milliseconds: 200),
      decoration: const BoxDecoration(
        borderRadius: BorderRadius.all(
          Radius.circular(10),
        ),
      ),
      height: 36,
      width: widget.width - 65,
      child: Theme(
        data: ThemeData(
          colorScheme: ColorScheme.fromSwatch().copyWith(secondary: Colors.transparent),
          primaryColor: Colors.white,
        ),
        child: TextField(
          focusNode: _focusNode,
          controller: widget.controller.textEditingController,
          onChanged: (value) {
            if (_isExpanded == false) _toggleExpand();
            setState(() {});
            widget.controller.filterItems(value);
            setState(() {});
            entry!.markNeedsBuild();
          },
          textAlignVertical: TextAlignVertical.center,
          style: widget.selectedTextStyle ?? const TextStyle(color: Colors.black),
          cursorColor: Colors.grey,
          decoration: InputDecoration(
            isCollapsed: true,
            floatingLabelBehavior: FloatingLabelBehavior.never,
            border: InputBorder.none,
            contentPadding: const EdgeInsets.only(left: 16, right: 16, top: 8, bottom: 12),
            labelText: widget.hintText,
            labelStyle: widget.hintTextStyle ?? const TextStyle(color: Colors.grey),
          ),
        ),
      ),
    );
  }
}

class _DropdownOverlay extends StatefulWidget {
  final VoidCallback onDismiss;

  const _DropdownOverlay({required this.onDismiss});

  @override
  State<_DropdownOverlay> createState() => __DropdownOverlayState();
}

class __DropdownOverlayState extends State<_DropdownOverlay> {
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onTap: widget.onDismiss,
      child: Material(
        color: Colors.transparent,
        child: Container(
          color: Colors.transparent,
          height: MediaQuery.of(context).size.height,
          width: MediaQuery.of(context).size.width,
        ),
      ),
    );
  }
}

class AnimatedSearchBoxController {
  final List<AnimatedDropdownItem> items;
  AnimatedDropdownItem? selectedValue;
  AnimatedDropdownItem initialValue;
  final List<AnimatedDropdownItem> filteredItems = [];
  final TextEditingController textEditingController = TextEditingController();

  AnimatedSearchBoxController({
    required this.items,
    required this.initialValue,
  })  : selectedValue = initialValue,
        assert(items.isNotEmpty);

  void filterItems(String keyword) {
    filteredItems.clear();
    if (keyword.isEmpty) {
      filteredItems.addAll(items);
    } else {
      filteredItems.addAll(items.where((item) => item.text.toLowerCase().contains(keyword.toLowerCase())));
    }
  }

  void initialize() {
    filteredItems.addAll(items);
  }
}

abstract class AnimatedDropdownItem {
  String get text;
  String get value;
}
